examples/vision/depth_estimation.py

```diff
@@ -0,0 +1,489 @@
+    u3 = UpscaleBlock(f[1])(u2, c2)
+    u4 = UpscaleBlock(f[0])(u3, c1)
+
+    outputs = layers.Conv2D(1, (1, 1), padding="same", activation="tanh")(u4)
+
+    model = tf.keras.Model(inputs, outputs)
+    return model
+
+
+"""
+## Optimizing Loss
+We have tried to optimize 3 losses in our model.
+1. Structural similarity index(SSIM).
+2. L1-loss, or Point-wise depth in our case.
+3. Edge wide depth with depth smoothness.
+
+Out of the three loss functions SSIM contributed the most in improving model performance.
+"""
+
+
+class DepthEstimationModel(tf.keras.Model):
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+
+    def compile(self, optimizer):
+        super(DepthEstimationModel, self).compile()
+        self.optimizer = optimizer
+        self.loss_metric = tf.keras.metrics.Mean(name="loss")
+
+    @property
+    def metrics(self):
+        return [self.loss_metric]
+
+    def train_step(self, batch_data):
+        input, target = batch_data
+        with tf.GradientTape() as tape:
+            pred = self.model(input, training=True)
+
+            # Edges
+            dy_true, dx_true = tf.image.image_gradients(target)
+            dy_pred, dx_pred = tf.image.image_gradients(pred)
+            weights_x = tf.exp(tf.reduce_mean(tf.abs(dx_true)))
+            weights_y = tf.exp(tf.reduce_mean(tf.abs(dy_true)))
+
+            # depth_smoothness
+            smoothness_x = dx_pred * weights_x
+            smoothness_y = dy_pred * weights_y
+
+            l_edges = tf.reduce_mean(abs(smoothness_x)) + tf.reduce_mean(
+                abs(smoothness_y)
+            )
+
+            # Structural similarity (SSIM) index
+            l_ssim = tf.reduce_mean(
+                1
+                - tf.image.ssim(
+                    target,
+                    pred,
+                    max_val=WIDTH,
+                    filter_size=7,
+                    k1=0.01 ** 2,
+                    k2=0.03 ** 2,
+                )
+            )
+            # Point-wise depth
+            l1_loss = tf.reduce_mean(tf.abs(target - pred))
+
+            # Weights
+            w1 = 0.85
+            w2 = 0.1
+            w3 = 1.0
+
+            loss = (w1 * l_ssim) + (w2 * l1_loss) + (w3 * l_edges)
+
+        gradients = tape.gradient(loss, self.model.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.model.trainable_variables))
+        self.loss_metric.update_state(loss)
+        return {
+            "Loss": self.loss_metric.result(),
+        }
+
+    def call(self, x):
+        pass
+
+
+"""
+## Model training
+"""
+
+optimizer = tf.keras.optimizers.Adam(
+    learning_rate=LR,
+    amsgrad=False,
+)
+model = unet_model(HEIGHT, WIDTH)
+DEM = DepthEstimationModel(model=model)
+DEM.compile(optimizer)
+
+train_loader = DataGenerator(
+    data=df[:600].reset_index(drop="true"), batch_size=BATCH_SIZE, dim=(HEIGHT, WIDTH)
+)
+DEM.fit(train_loader, epochs=EPOCHS)
```

Comment:
Please monitor performance on a validation set so you can avoid overfitting and have some visibility into generalization power